jetty-server/src/main/java/org/eclipse/jetty/server/Response.java

```
@@ -1305,6 +1305,29 @@ public void setTrailerFields(Supplier<Map<String, String>> trailers)
         return meta;
     }
 
+    /**
+     * Restore the CommittedMetaData, overwriting the current
+     * metadata in the process.
+     * <p>
+     *     This is useful when you want to see the state of
+     *     the Response object as it was when it was committed.
+     * </p>
+     */
+    public void restoreCommittedMetaData()
+    {
+        MetaData.Response committedResponse = getCommittedMetaData();
+        if (committedResponse != null)
+        {
+            // restore status
+            setStatus(committedResponse.getStatus());
+
+            // restore headers
+            HttpFields.Mutable responseFields = getHttpFields();
+            responseFields.clear();
+            responseFields.putAll(committedResponse.getFields());
```

I'm now worried about the performance impact of this.    On every response we are clearing any copying all the response headers just in case of the very VERY few cases that they are modified after commit.

I'm now not inclined to do this, at least not by default.

@lachlan-roberts  is there anyway the requestLogs could directly access the committed fields without the need to copy them back?

---

This is only done if a RequestLog is defined.

---

I felt the impact of having a switch on the various .getHeader() methods would be worse (performance wise).
As it would impact 100% of users, even those that don't use RequestLog.
WDYT?

---

As for "very few cases", like 100% of all webapps that use spring-security or use jersey?

---

Would it be possible to make the various .setHeader() implementations not actually set the header if the response is already committed?

---

There's potentially more waste (performance wise) of allowing the header to be set (even a new header) after response commit.

---

@gregw the request log uses the jetty `Request` and `Response` classes so it does have access to the committed metadata.

---

@lachlan-roberts but RequestLog implementations (like logback-access, spring-boot, jersey) have generic Request logging based on the HttpServletRequest and HttpServletResponse.

---

Is a check like ... `if (committedMetadata != metadata)` possible without worse performance?
ooh, how about a setting a "dirty" flag on the Response if any response header is set after commit?
We could use that to determine if we need to copy the headers back (or not).
That should cover the "very few cases" aspect, right?   (I think there might be a race condition tho)

---

A `committedMetadata.isEqualTo(metadata)` would fail fast, but be a full sweep for the most common case of no changes.
A dirty bit might be better, but still it would either have to be in HttpFields (and cost every other usage), or in Response, adding more complexity in more places.

So I'm probably more inclined to have an option in HttpChannel or HttpConfiguration for "restoreResponseFields", that defaults to off.

Or alternately, once committed, we could make the response headers immutable, so any calls would become noops?